### DISCUSSION TITLE
demo/c/positions.c: uses J2000 for RA/DEC but OF_DATE for horizontal coordinates

### DISCUSSION THREAD

**Discussion by KJ7NLL** [cosinekitty/astronomy#305] [General]
This is most likely a bug in _my_ understanding and _not_ a bug in the program, so perhaps you can help me understand what is going on here.  I have removed error checking for ease of reading:

```c
        equ_2000 = Astronomy_Equator(body[i], &time, observer, EQUATOR_J2000, ABERRATION);

        equ_ofdate = Astronomy_Equator(body[i], &time, observer, EQUATOR_OF_DATE, ABERRATION);

        hor = Astronomy_Horizon(&time, observer, equ_ofdate.ra, equ_ofdate.dec, REFRACTION_NORMAL);

        printf("%-8s %8.2lf %8.2lf %8.2lf %8.2lf\n", Astronomy_BodyName(body[i]), equ_2000.ra, equ_2000.dec, hor.azimuth, hor.altitude);
```

Note that `Astronomy_Horizon` uses `equ_ofdate.ra, equ_ofdate.dec` to calculate horizontal coordinates of the observer, but `printf` displays RA/DEC of the planet based on `equ_2000`.  

Assuming the code is correct, is it for some reason "more accurate" (or more "correct") to show planet RA/DEC positions in J2000 but still calculate horizontal coordinates based on from EQUATOR_OF_DATE?  

Thanks for your help!

**Comment by cosinekitty** [comment]
Hi @KJ7NLL and welcome to the project! I converted this from an Issue to a Discussion because it's a great question, and I'm sure this will help a lot of other people. Therefore, I want to leave it open for people to find.

Both kinds of RA/DEC, or equatorial coordinates, are equally accurate answers, but to two different questions.

We use J2000 equatorial coordinates when we are talking about locations in the sky that we want to remain valid over long spans of time. For example, if you asked where in the sky a particular star is, astronomers could use J2000 equatorial coordinates to answer you. That's because it is a fixed frame of reference that doesn't change over time. The same RA/DEC will be valid for a given star in 1950 as in 2050.

However, because the orientation of the Earth's rotation axis changes over time, we want to use "of-date" equatorial coordinates when that matters for a given question. Calculating horizontal coordinates for an observer on the Earth is such a situation. If you want to know how high a given star is above the horizon for a given observer at a given date and time, you need to know exactly how the Earth is oriented with respect to the rest of the universe, which means you need to use the Earth's axis at that time. So then we use of-date equatorial coordinates to give the actual position of the star relative to the Earth's axis and equator at that moment in time.

That `printf` above supposes that we want to report equatorial coordinates in the fixed J2000 system, such as what you would find in an almanac or ephemeris, along with horizontal coordinates as an observer would actually measure.

I hope this helps. Does this make sense?

**Reply by KJ7NLL** [reply]
Thanks for the explanation! I think that helps us understand for planets. What about for stars? For example:

If we have a J2000 coordinate for a star, and we want the horizon coordinate, do we need to use  `Astronomy_Equator`, or should we just pass RA/DEC to `Astronomy_Horizon` directly?

If we need to use `Astronomy_Equator` for a star, is it still a good idea to use `EQUATOR_OF_DATE` before passing it to `Astronomy_Horizon` for the best positional accuracy for our position?

We would like to get the best local accuracy because we are using stars to calibrate an azimuth/elevation rotor that we built to track satellites. We're going to attach a telescope to optically calibrate the antenna position. If the telescope can view the star in the center and we can get the AZ/ALT position, than we can use that to calibrate for talking to the International Space Station.

Here is a video about the project if you are interested in learning more:
 - [Project Presentation at CadenceLIVE 2023](https://youtu.be/vKeOOaMB4ZQ)
 - [Satellite Tracking Demo](https://youtu.be/vrlw4QPKMRY)

-Zeke

**Reply by cosinekitty** [reply]
Hi Zeke! Yes, if you have a J2000 coordinate for a star, you need to convert it to "of date" coordinates before passing it to `Astronomy_Horizon`, just like with a planet. And yes, you could do that using `Astronomy_Equator` after creating a user-defined star. See [Astronomy_DefineStar](https://github.com/cosinekitty/astronomy/tree/master/source/c#astronomy_definestarbody-ra-dec-distancelightyears--astro_status_t). This function allows you to treat an arbitrary point in the sky as a celestial body like any other, given its J2000 coordinates and distance in light years.

I'm looking forward to watching your videos this weekend! Thanks for sending them!

**Reply by KJ7NLL** [reply]
We might have just found an answer to the one of the questions above where your `Astronomy_Horizon` comment in `astronomy.c` says:
> _"The right ascension `ra` and declination `dec` passed in must be *equator of date* coordinates, based on the Earth's true equator at the date and time of the observation. Otherwise the resulting horizontal coordinates will be inaccurate."_

Great documentation.  If you have any other comments related to how we are calculating observer position for the purposes of getting the best horizontal coordinates then that would be great!

Here is our code for iterating the [brightest 200 stars](https://github.com/KJ7NLL/space-ham/blob/master/stars.c):

```c
for (i = 0; i < NUM_STARS; i++) 
{    
        idx++;

        if (n != idx
                && argc >= 3 
                && !strcasestr(stars[i].name, args[2]))
                continue;

        found++;
        found_star_idx = i; 

        Astronomy_DefineStar(BODY_STAR8,
                stars[i].ra, stars[i].dec, stars[i].dist_ly);

        equ_ofdate = Astronomy_Equator(BODY_STAR8, &time, observer, EQUATOR_OF_DATE, ABERRATION);
        if (equ_ofdate.status != ASTRO_SUCCESS)
        {    
                printf("%s: Astronomy_Equator returned status %d trying to get coordinates of date.\r\n",
                        stars[i].name,
                        equ_ofdate.status);
        }                    

        // Use the ra/dec from Astronomy_Equator(EQUATOR_OF_DATE)
        // for horizontal position to correct for light travel,                                          
        // parallax, and aberration---but show the star's
        // actual J2000 ra/dec for informational display.                                                
        hor = Astronomy_Horizon(&time, observer, equ_ofdate.ra, equ_ofdate.dec, REFRACTION_NORMAL);      
        printf("%3d. %-37s %8.2lf %8.2lf %8.2lf %8.2lf %8.2lf\r\n",                                      
               idx, stars[i].name,                                                                       
               stars[i].ra, //equ_ofdate.ra,                                                             
               stars[i].dec,// equ_ofdate.dec,                                                           
               hor.azimuth, hor.altitude,                                                                
               stars[i].mag_vis);                                                                        
}                                                                                                        

```

**Reply by KJ7NLL** [reply]
(It looks like you answered at the same time I replied on the previous message, thanks for confirming!  Feel free to comment on our code if you see room for improvement.)

**Reply by cosinekitty** [reply]
That code is how I would have done it too. I'll be very interested to see how this satellite tracking project evolves. It looks like it could be an ideal case study I would want to link to, especially if you are willing to keep your code and other media open source.

**Reply by KJ7NLL** [reply]
Our code is open source, here it is on github if you would like to check it out: https://github.com/KJ7NLL/space-ham

This is the commit where we added the ability to track celestial bodies: https://github.com/KJ7NLL/space-ham/commit/a20b66859d67c4c97d9cd3641404374b6f0f73d6

It is pretty neat that your code is so accurate with such a small code footprint. We have this working on a 32-bit micro-controller with 96k of RAM at 34 MHz!

Now we can run commands like `astro track venus` and `astro track sirius`:

```py
[Zeke&Daddy@console]# astro track sirius
  n. BODY                                        RA      DEC       AZ      ALT      MAG
 11. Alpha Canis Majoris (Sirius)              6.75   -16.70   263.86   -17.13    -1.44
UTC: 06/05/2023  04:37:26  unix time: 1685939846
Uptime: 0.27 hours - Your lat/long: 45.000000 N -122.000000 E; altitude: 0.076200
iadc[0]: 1.360293 volts
iadc[1]: 2.126667 volts
iadc[2]: 0.635824 volts
iadc[3]: 0.000000 volts

theta cal: [0.00, 720.00] deg = [1.73, 1.07] volts, -0.9077 mV/deg
phi cal: [-90.00, 90.00] deg = [0.48, 2.67] volts, 12.1819 mV/deg
rotor2 cal: [0.00, 0.00] deg = [0.00, 0.00] volts, nan mV/deg
focus cal: [0.00, 0.00] deg = [0.00, 0.00] volts, nan mV/deg

theta pos=404.83 deg, target=404.83 deg [online, port=B0/B1]: P=0.0% I=0.0% D=-0.0% speed=0.0%
phi pos=45.21 deg, target=45.21 deg [online, port=C0/C1]: P=0.0% I=0.0% D=-0.0% speed=0.0%
rotor2 pos=0.00 deg, target=0.00 deg [invalid, port=?0/?0]: P=0.0% I=0.0% D=0.0% speed=0.0%
focus pos=0.00 deg, target=0.00 deg [invalid, port=C2/C3]: P=0.0% I=0.0% D=0.0% speed=0.0%
No satellite is being tracked
Tracking celestial body: Alpha Canis Majoris (Sirius)
```

**Reply by KJ7NLL** [reply]
Hello @cosinekitty,

I made a video where we track the moon using Astronomy Engine. Here is the link: https://youtu.be/NMwvoyL542Y?si=fNVn9N9q3S_RDKTm

Enjoy!
